server/src/main/java/org/elasticsearch/index/codec/vectors/IVFVectorsReader.java

```diff
@@ -279,7 +283,10 @@ public final void search(String field, float[] target, KnnCollector knnCollector
         long expectedDocs = 0;
         long actualDocs = 0;
         // initially we visit only the "centroids to search"
-        while (centroidQueue.size() > 0 && centroidsVisited < nProbe && actualDocs < knnCollector.k()) {
+        // Note, numCollected is doing the bare minimum here.
+        // TODO do we need to handle nested doc counts similarly to how we handle
+        // filtering? E.g. keep exploring until we hit an expected number of parent documents vs. child vectors?
+        while (centroidQueue.size() > 0 && centroidsVisited < nProbe && knnCollectorImpl.numCollected() < knnCollector.k()) {
```

Comment:
I did consider doing something similar to our filtering logic, by treating the number of visited vectors, vs. the number of visited parent docs, but I am not 100% sure its absolutely necessary. 

If it is necessary, we will need to add some bit set logic to the collector to keep track of the visited parent docs as we cannot do a simple incremental count as we might visit the same parent document multiple times.

Reply:
If we want to keep going until we collected k, shouldn't the condition be:

 `centroidQueue.size() > 0 && (centroidsVisited < nProbe| | knnCollectorImpl.numCollected() < knnCollector.k()))`